### PR TITLE
utils: implement our own lrint() for Visual C++

### PR DESCRIPTION
**Pull request by xenu**
Commit cd35a4dd added the -fno-math-errno flag which allows lrint() to be inlined on GCC and Clang.

Unfortunately, Visual C++ can't inline lrint() without -fp:fast. That flag has some nasty implications, so I think it's better to just write our own lrint.

This change helps a bit with the video sample from #736. It halves the number of dropped frames in MPC-HC on my machine (from ~180 to ~80).

### CODE REVIEW

**Comment by xenu** [comment]
Compiler Explorer proof that the generated machine code of ass_lrint() is identical to how GCC inlines lrint(): https://godbolt.org/z/aPWcGqjrE

**Comment by NBruderman** [comment]
@astiob is looks like this quite help with my issue. Do you see any downsides to this?

File: libass/ass_utils.h
Comment (TheOneric) [inline libass/ass_utils.h:175]: This adds a function call indirection on all platforms which we previously got rid off with `-fno-errno-math`. This is not acceptable. Instead `ass_lrint` should be a macro expanding to `lrint` on most platforms, or perhaps directly override `lrint` with a macro on MSVC

**Review by TheOneric — DISMISSED** [review]
Did you/someone ask Microsoft for a fine grained `no-errno-math` toggle yet?  

In general I’m not too fond of having too much platform-specific optimisations in the code and would much prefer a simple compiler toggle in configuration logic. This particular case may be relevant enough to add a code variant anyway, but on the other hand many more projects might profit from MSVC having a `no-errno-math` equivalent

File: libass/ass_utils.h
Comment (rcombs) [reply libass/ass_utils.h:175]: It's a `static inline`, and no function call is emitted in practice (though I wouldn't object to `#define ass_lrint lrint` on non-MSVC): https://godbolt.org/z/Pcn7TMddo

Defining macros over standard library symbols already provided by the libc is generally UB.

File: libass/ass_utils.h
Comment (rcombs) [inline libass/ass_utils.h:168]: Why the explicit exclusion of arm64 here? Shouldn't the X64/X86 conditions not pass on ARM?

File: libass/ass_utils.h
Comment (TheOneric) [reply libass/ass_utils.h:175]: oh, it’s in the header not `.c`. Yeah, it’s unlikely to end up as a function call then

**Review by rcombs — COMMENTED** [review]
Out of curiosity, why are you building with MSVC? (I'm trying to understand the use-cases behind this better.)

File: libass/ass_utils.h
Comment (xenu) [reply libass/ass_utils.h:168]: When compiling for the ARM64EC ABI the `_M_X64` macro is also defined, as [documented here](https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170).

The longer explanation is that there are two ABIs on ARM64 Windows. "ARM64" is the normal ABI and it behaves as one would expect. "ARM64EC" is a special ABI that tries to be as X64 compatible as possible to allow linking with emulated X64 binaries.

**Comment by xenu** [comment]
>Out of curiosity, why are you building with MSVC? (I'm trying to understand the use-cases behind this better.)

Because that's what MPC-HC does.

**Comment by xenu** [comment]
>In general I’m not too fond of having too much platform-specific optimisations in the code and would much prefer a simple compiler toggle in configuration logic. This particular case may be relevant enough to add a code variant anyway, but on the other hand many more projects might profit from MSVC having a no-errno-math equivalent

I generally agree with your concern, but I would argue that this particular case is not *that* invasive, and the performance benefit is very noticeable.

File: libass/ass_utils.h
Comment (rcombs) [reply libass/ass_utils.h:168]: Huh! Bizarre, but I suppose it's harmless. I see that MSVC ends up emitting a call to `_CVTSD2SI32` here, and I wonder how that differs from `lrint`.

**Comment by MrSmile** [comment]
I feel that `/fp:fast` is a much better workaround for this. Should be fairly safe especially in MSVC, as it's not that aggressive at optimizing as GCC and clang.

**Comment by xenu** [comment]
In both `-fp:fast` and `-ffast-math`, I'm always uncomfortable with associative math optimizations, as they have the potential to change the results. Of course, it's very likely that it doesn't affect libass. I'm not really familiar with this code, certainly not as much as you guys.

BTW, while researching for this PR, I found [this ticket](https://developercommunity.visualstudio.com/t/MSVCs-sincos-implementation-is-incorrec/10582378) about a nasty bug in a combined sin + cos optimization in `/fp:fast`. I found this because I noticed this pattern in libass and I wondered if MSVC has sincos (it doesn't). That said, it seems this bug mostly affects very large numbers so I think it shouldn't cause problems in practice.

**Comment by xenu** [comment]
PS. I've built unmodified libass with `/fp:fast` and the number of dropped frames is about the same as with my patch. It seems that inlining lrint() is the only FP optimization that significantly affects performance in that sample video.

**Comment by TheOneric** [comment]
> (In my experience Microsoft tends to ignore feature requests.)

I didn’t find anything explicitly requesting an `-fno-errno-math` flag on the page you linked earlier. But there’s https://developercommunity.visualstudio.com/t/trivial-math-functions-as-rintf-generate/10469611 complaining about `rintf` etc not compiling to a simple instructions and it was marked as fixed recently, supposedly available in the preview build since yesterday.
Iiuc `rintf` should also sometimes set `errno`, e.g. when rounding leads to `HUGE_VAL`, but perhaps they are able to determine whether errno get read or not.

Could you check if this already helps? If not we probably want to carry this code-tweak (for some while) due to its large impact, but i still think it’d be wort to at least try to file a feature request

**Comment by xenu** [comment]
>Could you check if this already helps? 

I've installed the latest preview (17.11 Preview 5) and none of the examples provided by the author of that ticket are inlined, with or without `/fp:fast`. I don't understand this.

There's no change with `lrint`, it's inlined only with `/fp:fast`.

**Comment by xenu** [comment]
>This affects clang code generation for msvc target, so please explicitly exclude clang via __clang__.

Done (although it's a bit pedantic as the generated code should be the same).

**Comment by xenu** [comment]
>but i still think it’d be wort to at least try to file a feature request

https://developercommunity.visualstudio.com/t/Add-a-flag-similar-to--fno-errno-math-th/10710444

**Comment by astiob** [comment]
> This affects clang code generation for msvc target, so please explicitly exclude clang via `__clang__`.

If Clang is being used in MSVC emulation mode, doesn’t it _not_ get `-fno-errno-math`, therefore doesn’t it fail to inline `lrint`, and therefore don’t we _want_ to include it in this optimization?

**Comment by xenu** [comment]
It's... complicated. Clang can be configured to use the MinGW toolchain or to use the Visual C++ toolchain.

It also has two drivers, both are present in either configuration.

`clang.exe` sets `_MSC_VER` if it's configured to use the VC++ toolchain. But it takes standard flags regardless of the configuration, so `-fno-math-errno` works there.

`clang-cl.exe` always sets `_MSC_VER` regardless of the configuration and since its commandline is compatible with VC++, it doesn't understand `-fno-math-errno`.

It's a mess.

I imagine Meson users will typically use the `clang.exe` frontend.

**Comment by astiob** [comment]
> so I used clang trunk and found that intrinsics generate suboptimal code.

Out of curiosity, what was suboptimal about the code? Should we be worried and search for alternatives without the intrinsic?

**Comment by xenu** [comment]
>Out of curiosity, what was suboptimal about the code? Should we be worried and search for alternatives without the intrinsic?

I have a suspicion what happened:

ass_lrint() assumes that `sizeof(long) == 4` which is [always true on Windows](https://learn.microsoft.com/en-us/cpp/cpp/data-type-ranges?view=msvc-170) on all architectures.

When you switch to Clang in the Compiler Explorer, you switch to Linux, so the size of `long` changes to 8. Because of that, Clang emits additional `cdqe` instruction (convert doubleword to quadword). It disappears if you switch the function's return type to `int` or switch the intrinsic to the 64-bit version.

@Andarwinux is this what happened?

**Review by TheOneric — APPROVED** [review]

**Comment by NBruderman** [comment]
Any news about this merge? Or is this on hold until it's time for the next release?

**Comment by kasper93** [comment]
`lrint` is not the only one affected function. Generally I think mid-term goal should be to adjust the code to run safely with `/fp:fast` as this is most likely fine for libass needs. Only few NaN checks would need to be handled differently, but not biggie.

Reimlementing standard library functions is not feasible approach long-term.

**Comment by xenu** [comment]
> Only few NaN checks would need to be handled differently, but not biggie.

NaN behaviour is far from the only difference between `/fp:strict` (the default) and `/fp:fast`.

Anyway, I'm abandoning this PR because this is getting tiring.

**Comment by NBruderman** [comment]
> > Only few NaN checks would need to be handled differently, but not biggie.
> 
> NaN behaviour is far from the only difference between `/fp:strict` (the default) and `/fp:fast`.
> 
> Anyway, I'm abandoning this PR because this is getting tiring.

Why closing this? Just keep it open until the maintainers will decide what approach to take. You don't need to work on this anymore, just keep it open

**Comment by kasper93** [comment]
> Anyway, I'm abandoning this PR because this is getting tiring.

Opened last week... Approved last week... `+41 −26`... tiring. Ok, buddy. 

Back on topic.

> NaN behaviour is far from the only difference between /fp:strict (the default) and /fp:fast.

Of course, but I argue that only NaN is affecting current libass code base in a meaningful way, because there is two checks that would be broken by that. I've analyzed it last time, with similar discussion about `-fno-math-errno` and siblings. The rest of code does not require IEEE compliant floating point operations.

**Comment by xenu** [comment]
>PR here can take quite a while to be merged.

Waiting isn't an issue, I've had PRs that waited for years in other projects.

But I'm not invested in this project enough to continue arguing in this thread, especially with people like this guy above. This is the part that is tiring.

**Comment by NBruderman** [comment]
Maybe you can open this PR just so people remember it exists, while the discussion can take place here?
https://github.com/libass/libass/issues/782
It feels like just too important to be just left out. It's basically 2.2x performance gain.

**Comment by astiob** [comment]
Merged as https://github.com/libass/libass/commit/b13f5d107237abb15487dc031f83c11e369a4be9. Thanks! We can improve on this later if/when we decide on a better way to handle this (in #782 or otherwise).

(I can’t say that I, personally, fully understand why this was closed and deleted, but everyone here is a volunteer. If the discussion seemed to become unpleasant, I feel sorry; but I can respect the wish to abandon it. Thanks for your contribution!)

**Pull request closed without merging**
